File: 2D_DynamicProgramming/coinChange2/coinChange2.cpp
#include<bits/stdc++.h>
using namespace std;
 
#define ll long long
#define mod 1000000007

// Dynamic Programming Unbounded Knapsack - Leetcode 518
/*
    Given array of coins & an amount, return # of combos that make up this amount
    Ex. amount = 5, coins = [1,2,5] -> 4 (5, 2+2+1, 2+1+1+1, 1+1+1+1+1)

    DFS + memo, 2 choices: either try coin & stay at idx, or don't try & proceed

    Time: O(m x n)
    Space: O(m x n)
*/

//         5   4   3   2   1   0      <-amount
//     __________________________
// c 1  |  4   3   2   2   1    1
// o    |------------------------                       1
// i 2  |  1   1   0   1   0    1
// n    |------------------------
// s 5  |  1   0   0   0   0    1
// |
// v
// Any cell value-> indicates whats the numOfWays to create amount(col) when you have coins (row, row +1.. etc)
// 
class Solution_TD {
public:
    int change(int amount, vector<int>& coins) {
        return dfs(amount, coins, 0, 0);
    }
private:
    // {(index, sum) -> # of combos that make up this amount}
    map<pair<int, int>, int> dp;
    
    int dfs(int amount, vector<int>& coins, int i, int sum) {
        if (sum == amount) {
            return 1;
        }
        if (sum > amount) {
            return 0;            // overshooting returns 0
        }
        if (i == coins.size()) {
            return 0;            //   
        }
        if (dp.find({i, sum}) != dp.end()) {
            return dp[{i, sum}];
        }
        
        dp[{i, sum}] = dfs(amount, coins, i, sum + coins[i]) // taking current coin
                     + dfs(amount, coins, i + 1, sum);       // ignore current coin
        
        return dp[{i, sum}];
    }
}; 

class Solution_BU {
public:
    int change(int amount, vector<int>& coins) {
        int n = coins.size();
        // no. ways
        vector<vector<int> > dp(amount+1, vector<int>(coins.size()+1, 0) );
        for(int a=0;a<=amount;a++){
            dp[a][n] = 0;           // with no coin to chosse, no amount can be formed
        }
        for(int c=0;c<=n;c++){
            dp[0][c] = 1;       // amount 0 can be made up in 1 way by not selecting the coin
        }
        
        for(int a=1;a<=amount;a++){
            for(int c=n-1;c>=0;c--){  // selecting coins from index c till n-1
                int val = coins[c];
                //1-way ignoring current coin
                dp[a][c] = dp[a][c+1];
                if((a-val) >= 0)
                    dp[a][c] += dp[a-val][c]; 
            }
        }
        return dp[amount][0];
    }
};
int main(){
    
    return 0;
}

File: 2D_DynamicProgramming/coinChange2/coinChange2.py
from typing import List
# Dynamic Programming Unbounded Knapsack - Leetcode 518
#         5   4   3   2   1   0      <-amount
#     __________________________
# c 1  |  4   3   2   2   1    1
# o    |------------------------                       1
# i 2  |  1   1   0   1   0    1
# n    |------------------------
# s 5  |  1   0   0   0   0    1
# |
# v
# Any cell value-> indicates whats the numOfWays to create amount(col) when you have coins (row, row +1.. etc)
# 
class Solution:
    def change_rec(self, amount: int, coins: List[int]) -> int:
        # MEMOIZATION
        # Time: O(n*m)
        # Memory: O(n*m)
        cache = {}  # <coin_index, amount> => # of combos that make up this amount

        def dfs(i, a):
            if a == amount:
                return 1
            if a > amount:
                return 0
            if i == len(coins):
                return 0
            if (i, a) in cache:
                return cache[(i, a)]
            # walk down the rec tree
            cache[(i, a)] = dfs(i, a + coins[i]) + dfs(i + 1, a) # choose curr coin + ignore curr coin
            return cache[(i, a)]

        return dfs(0, 0)

        # DYNAMIC PROGRAMMING
        # Time: O(n*m)
        # Memory: O(n*m) => 2d grid of  amount X coins
        # filling grid amount X coins each col at a time
        # dp[a][i] => number of ways of getting amount 'a' with coins from ith index to n-1 index
    def change_DP(self, amount: int, coins: List[int]) -> int:    
        dp = [[0] * (len(coins) + 1) for i in range(amount + 1)] # <= 2D  grid amount X coins 
        dp[0] = [1] * (len(coins) + 1) # amount 0 can be obtained by 1 way for each coin i.e by not selecting it
        for a in range(1, amount + 1): # building dp from small amt to given amount
            for i in range(len(coins) - 1, -1, -1):
                dp[a][i] = dp[a][i + 1] # not selecting coin 'i'
                if a - coins[i] >= 0:
                    dp[a][i] += dp[a - coins[i]][i]  # selecting coin 'i' if it fits within current amount 'a'
        return dp[amount][0] # no. of ways of obtaining 'amount' with all given coins

        # DYNAMIC PROGRAMMING
        # Time: O(n*m)
        # Memory: O(n) where n = amount
        #  dp     [ 5   4   3   2   1   0 ]     <-amount
        #                            1
        # nextDP  [ 5   4   3   2   1   0 ]     <-amount
        #                            1
    def change_DP_opt(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1
        for i in range(len(coins) - 1, -1, -1): # looping over coins first
            nextDP = [0] * (amount + 1)
            nextDP[0] = 1

            for a in range(1, amount + 1):
                nextDP[a] = dp[a] # equivalent to 'ignoring current coin' when you have current and all prev processed coins
                if a - coins[i] >= 0:
                    nextDP[a] += nextDP[a - coins[i]] # taking current coin
            dp = nextDP
        return dp[amount]